File: typescript/packages/mastra/src/tools.ts
import { createTool } from '@mastra/core/tools';
import { z } from 'zod';
import {
  DISCOVER_DESCRIPTION,
  DISCOVER_QUERY_PARAM,
  INTERPRET_DESCRIPTION,
  INTERPRET_ITEMS_PARAM,
  INTERPRET_QUERY_PARAM,
  SEARCH_DESCRIPTION,
  SEARCH_QUERY_PARAM,
} from '@copass/config';
import type { CopassClient, SearchPreset, WindowLike } from '@copass/core';

export interface CopassToolsOptions {
  /** An authenticated `@copass/core` client. */
  client: CopassClient;
  /** Sandbox all retrieval runs against. */
  sandbox_id: string;
  /** Optional project scoping for retrieval calls. */
  project_id?: string;
  /**
   * Optional conversation window — a {@link WindowLike} (typically a
   * `ContextWindow` from `client.contextWindow.create()`). When provided,
   * every retrieval call is automatically window-aware.
   */
  window?: WindowLike;
  /**
   * Preset for `discover`, `interpret`, and `search`. Defaults to
   * `"copass/copass_1.0"`. Under `"copass/copass_2.0"` discover items
   * carry `subgraph` (pre-rendered ASCII tree) and `matched_query_nodes`
   * fields. Append `":thinking"` (e.g. `"copass/copass_2.0:thinking"`)
   * to enable task decomposition before retrieval on `search`.
   */
  preset?: SearchPreset;
}

/**
 * Return Copass retrieval as a set of Mastra tool objects (built with
 * `createTool` from `@mastra/core/tools`).
 *
 * Agent-framework-neutral shape: the LLM chooses whether to surface a menu
 * (`discover`), drill into specific items (`interpret`), or get a
 * synthesized answer in one shot (`search`).
 *
 * @example
 * ```ts
 * import { Agent } from '@mastra/core/agent';
 * import { anthropic } from '@ai-sdk/anthropic';
 * import { copassTools } from '@copass/mastra';
 *
 * const window = await copass.contextWindow.create({ sandbox_id });
 * const tools = copassTools({ client: copass, sandbox_id, window });
 *
 * const agent = new Agent({
 *   name: 'support-bot',
 *   instructions: 'Answer questions using the knowledge graph.',
 *   model: anthropic('claude-opus-4-7'),
 *   tools,
 * });
 *
 * const response = await agent.generate('why is checkout flaky?');
 * ```
 */
export function copassTools(options: CopassToolsOptions) {
  const { client, sandbox_id, project_id, window, preset = 'copass/copass_1.0' } = options;

  const discover = createTool({
    id: 'discover',
    description: DISCOVER_DESCRIPTION,
    inputSchema: z.object({
      query: z.string().describe(DISCOVER_QUERY_PARAM),
    }),
    execute: async ({ query }) => {
      const response = await client.retrieval.discover(sandbox_id, {
        query,
        project_id,
        window,
        preset,
      });
      return {
        header: response.header,
        // Project the v2 fields (`subgraph` + `matched_query_nodes`)
        // alongside the v1 fields. Populated only under
        // `copass/copass_2.0` (or its `copass/2.0` alias); `null` under
        // v1 — agents can ignore them when not present.
        items: response.items.map((item) => ({
          score: item.score,
          summary: item.summary,
          canonical_ids: item.canonical_ids,
          subgraph: item.subgraph ?? null,
          matched_query_nodes: item.matched_query_nodes ?? null,
        })),
        next_steps: response.next_steps,
      };
    },
  });

  const interpret = createTool({
    id: 'interpret',
    description: INTERPRET_DESCRIPTION,
    inputSchema: z.object({
      query: z.string().describe(INTERPRET_QUERY_PARAM),
      items: z.array(z.array(z.string())).min(1).describe(INTERPRET_ITEMS_PARAM),
    }),
    execute: async ({ query, items }) => {
      const response = await client.retrieval.interpret(sandbox_id, {
        query,
        items,
        project_id,
        window,
        preset,
      });
      return { brief: response.brief };
    },
  });

  const search = createTool({
    id: 'search',
    description: SEARCH_DESCRIPTION,
    inputSchema: z.object({
      query: z.string().describe(SEARCH_QUERY_PARAM),
    }),
    execute: async ({ query }) => {
      const response = await client.retrieval.search(sandbox_id, {
        query,
        project_id,
        window,
        preset,
      });
      return { answer: response.answer };
    },
  });

  return { discover, interpret, search };
}

File: python/copass-langchain/src/copass_langchain/tools.py
"""LangChain tool adapters for Copass retrieval.

Python mirror of ``typescript/packages/langchain/src/tools.ts`` — uses
``StructuredTool.from_function`` (the LangChain-python equivalent of
the TS ``tool()`` factory) so tool names + descriptions come
programmatically from ``copass-config``.

The three returned tools — ``discover``, ``interpret``, ``search`` —
are window-aware when a ``window`` is supplied: the server will
exclude items already surfaced earlier in the conversation. Keeps
repeated ``discover`` calls cheap and productive.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, List, Optional

from copass_config import (
    DISCOVER_DESCRIPTION,
    DISCOVER_QUERY_PARAM,
    INTERPRET_DESCRIPTION,
    INTERPRET_ITEMS_PARAM,
    INTERPRET_QUERY_PARAM,
    SEARCH_DESCRIPTION,
    SEARCH_QUERY_PARAM,
)
from copass_core import CopassClient, SearchPreset
from langchain_core.tools import StructuredTool
from pydantic import BaseModel, Field

from copass_langchain.types import ContextWindowLike


@dataclass(frozen=True)
class CopassTools:
    """Bundle of the three Copass retrieval tools as
    ``StructuredTool`` instances. Pass ``tools.discover``,
    ``tools.interpret``, ``tools.search`` (or ``tools.all()``) to the
    agent framework."""

    discover: StructuredTool
    interpret: StructuredTool
    search: StructuredTool

    def all(self) -> List[StructuredTool]:
        return [self.discover, self.interpret, self.search]


class _DiscoverArgs(BaseModel):
    query: str = Field(..., description=DISCOVER_QUERY_PARAM)


class _InterpretArgs(BaseModel):
    query: str = Field(..., description=INTERPRET_QUERY_PARAM)
    items: List[List[str]] = Field(
        ...,
        min_length=1,
        description=INTERPRET_ITEMS_PARAM,
    )


class _SearchArgs(BaseModel):
    query: str = Field(..., description=SEARCH_QUERY_PARAM)


def copass_tools(
    *,
    client: CopassClient,
    sandbox_id: str,
    project_id: Optional[str] = None,
    window: Optional[ContextWindowLike] = None,
    preset: SearchPreset = "copass/copass_1.0",
) -> CopassTools:
    """Return a :class:`CopassTools` bundle wired to ``client``.

    Args:
        client: Authenticated :class:`CopassClient`.
        sandbox_id: Sandbox every retrieval call runs against.
        project_id: Optional project narrowing.
        window: Optional :class:`ContextWindowLike` — when provided,
            every retrieval call is window-aware.
        preset: Retrieval preset for ``discover`` / ``interpret`` /
            ``search``. Defaults to ``"copass/copass_1.0"``. Under
            ``"copass/copass_2.0"`` discover items carry ``subgraph``
            (pre-rendered ASCII tree) and ``matched_query_nodes``
            fields. Append ``":thinking"`` (e.g.
            ``"copass/copass_2.0:thinking"``) to enable task
            decomposition before retrieval on ``search``.

    Example::

        from copass_core import CopassClient, ApiKeyAuth
        from copass_langchain import copass_tools

        client = CopassClient(auth=ApiKeyAuth(key="olk_..."))
        tools = copass_tools(client=client, sandbox_id="sb_...")
        # Hand `tools.all()` to LangGraph's create_react_agent.
    """

    async def _discover(query: str) -> Dict[str, Any]:
        response = await client.retrieval.discover(
            sandbox_id=sandbox_id,
            query=query,
            project_id=project_id,
            window=window,
            preset=preset,
        )
        return {
            "header": response.get("header", ""),
            "items": [
                # Project the v2 fields (``subgraph`` + ``matched_query_nodes``)
                # alongside the v1 fields. Populated only under
                # ``copass/copass_2.0`` (or its ``copass/2.0`` alias);
                # ``None`` under v1.
                {
                    "score": item.get("score"),
                    "summary": item.get("summary"),
                    "canonical_ids": item.get("canonical_ids", []),
                    "subgraph": item.get("subgraph"),
                    "matched_query_nodes": item.get("matched_query_nodes"),
                }
                for item in response.get("items", [])
            ],
            "next_steps": response.get("next_steps", ""),
        }

    async def _interpret(query: str, items: List[List[str]]) -> Dict[str, Any]:
        response = await client.retrieval.interpret(
            sandbox_id=sandbox_id,
            query=query,
            items=items,
            project_id=project_id,
            window=window,
            preset=preset,
        )
        return {"brief": response.get("brief", "")}

    async def _search(query: str) -> Dict[str, Any]:
        response = await client.retrieval.search(
            sandbox_id=sandbox_id,
            query=query,
            project_id=project_id,
            window=window,
            preset=preset,
        )
        return {"answer": response.get("answer", "")}

    return CopassTools(
        discover=StructuredTool.from_function(
            coroutine=_discover,
            name="discover",
            description=DISCOVER_DESCRIPTION,
            args_schema=_DiscoverArgs,
        ),
        interpret=StructuredTool.from_function(
            coroutine=_interpret,
            name="interpret",
            description=INTERPRET_DESCRIPTION,
            args_schema=_InterpretArgs,
        ),
        search=StructuredTool.from_function(
            coroutine=_search,
            name="search",
            description=SEARCH_DESCRIPTION,
            args_schema=_SearchArgs,
        ),
    )


__all__ = ["CopassTools", "copass_tools"]
